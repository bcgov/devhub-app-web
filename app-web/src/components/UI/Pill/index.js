import React from 'react';
import PropTypes from 'prop-types';
import { faTimes, faTimesCircle } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import styled from '@emotion/styled';

const variants = {
  outlined: 'outlined',
  filled: 'filled',
};

const Container = styled.button`
  padding: 3px;
  border-radius: 1em;
  color: rgba(0, 0, 0, 0.87);
  margin: 4px 5px;
  position: relative;
  display: flex;
  align-items: center;
  border-color: transparent;
  border-width: 1px;
  border-style: solid;
  :focus {
    outline: none;
  }
  > span {
    margin: 0 4px;
    font-size: 0.875em;
  }
`;

const FilledContainer = styled(Container)`
  background-color: #e0e0e0;
`;

const OutlinedContainer = styled(Container)`
  border: 1px solid #ccc;
`;

const PillIcon = styled(FontAwesomeIcon)`
  margin: ${props => (props.icon.iconName === 'times' ? '0 6.75px' : '0 4px')};
`;
/**
 * Pill Component
 * @param {Object} props
 * @param {Object} props.label the text inside the pill
 * @param {Object} props.icon *optional* a react compnent that will render to the left of the text
 * @param {Object} props.variant *defaults to filled* the render style [outlined, filled]
 * @param {Object} props.onDelete called if deletable is true, receives label as an arg
 * @param {Object} props.onClick click handler for entire pill if passed in
 */
export const Pill = ({ label, icon, variant, onDelete, onClick, deletable, ...rest }) => {
  const props = {};

  if (onClick) {
    props.onClick = onClick;
  }

  if (variant === variants.filled) {
    return (
      <FilledContainer {...props} {...rest} data-variant={variant.filled}>
        <span>{label}</span>
        {deletable && (
          <PillIcon aria-label="delete" icon={faTimesCircle} onClick={() => onDelete(label)} />
        )}
      </FilledContainer>
    );
  } else if (variant === variants.outlined) {
    return (
      <OutlinedContainer {...props} {...rest} data-variant={variant.outlined}>
        <span>{label}</span>
        {deletable && (
          <PillIcon aria-label="delete" icon={faTimes} onClick={() => onDelete(label)} />
        )}
      </OutlinedContainer>
    );
  }
  return (
    <Container {...props} {...rest}>
      <span>{label}</span>
      {deletable && <PillIcon aria-label="delete" icon={faTimes} onClick={() => onDelete(label)} />}
    </Container>
  );
};

Pill.propTypes = {
  label: PropTypes.string.isRequired,
  icon: PropTypes.node,
  variant: PropTypes.oneOf(Object.keys(variants)),
  onDelete: PropTypes.func,
  onClick: PropTypes.func,
  deletable: PropTypes.bool.isRequired,
};

Pill.defaultProps = {
  variant: variants.filled,
  deletable: true,
};

export default Pill;
