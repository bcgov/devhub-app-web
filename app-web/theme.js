/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/

/*
Emotion theming object

We are transitioning from common elements such as colors, background colors being abstracted away into
a themes provider. This is a very common practice now a days and makes it maintain themes on the fly
docs on theming is here https://emotion.sh/docs/theming#___gatsby

Docs on emotion https://emotion.sh/docs/introduction

--- our idea is to introduce emotion early and slowly replace css modules with emotion styled components
when and as needed. This is a slow processes but will lead to better and more maintaible styling in the long run.
*/

import { SPACING } from './src/constants/designTokens';
import { RESOURCE_TYPES, TOPICS, SEARCH_RESOURCE_TYPES, JOURNEY } from './src/constants/ui';

export const DEVHUB_PALETTE = {
  [RESOURCE_TYPES.COMPONENTS]: '#4299D2',
  [RESOURCE_TYPES.DOCUMENTATION]: '#246BD1',
  [RESOURCE_TYPES.PEOPLE]: '#E17039',
  [RESOURCE_TYPES.SELF_SERVICE_TOOLS]: '#F3BA45',
  [TOPICS]: '#444',
  [RESOURCE_TYPES.REPOSITORIES]: '#6e5494',
  [SEARCH_RESOURCE_TYPES.GITHUB_ISSUE]: '#6e5494',
  [RESOURCE_TYPES.EVENTS]: '#f44b42',
  [JOURNEY]: '#009688',
  lightgreen: '#66bb6a',
  yellow: '#FCBA19',
  lightyellow: '#F3BA45',
  grey: '#494949',
  lightgrey: '#F2F2F2',
  darkgrey: '#444',
  orange: '#E17039',
  purple: '#6E5494',
  red: '#D8292F',
  lightred: '#F44B42',
  green: '#2E8540',
  blue: '#003366',
  babyblue: '#4299D2',
  lightblue: '#246BD1',
  linkblue: '#1A5A96',
  bgBlue: '#38598A',
  white: '#ffffff',
};

export default {
  colors: {
    ...DEVHUB_PALETTE,
  },
  padding: {
    main: {
      mobile: `${SPACING['2x']} ${SPACING['2x']}`,
      desktop: `${SPACING['2x']} ${SPACING['13x']}`,
    },
  },
  breakpoints: {
    // specific breakpoints
    main: {
      desktop: '@media screen and (min-width: 932px)',
    },
  },
};
