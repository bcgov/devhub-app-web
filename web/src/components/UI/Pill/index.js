import React from 'react';
import PropTypes from 'prop-types';
import {
  faTimes,
  faTimesCircle,
  faLayerGroup,
  faUser,
  faTag,
  faSearch,
  faProjectDiagram,
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import styled from '@emotion/styled';
import ResourceTypeIcon from '../ResourceTypeIcon';

const variants = {
  outlined: 'outlined',
  filled: 'filled',
};

const Container = styled.button`
  padding: 3px;
  border-radius: 1em;
  color: rgba(0, 0, 0, 0.87);
  margin: 4px 5px;
  position: relative;
  display: flex;
  align-items: center;
  border-color: transparent;
  border-width: 1px;
  border-style: solid;
  :focus {
    outline: none;
  }
  > span {
    margin: 0 4px;
    font-size: 0.875em;
  }
  @media only screen and (max-width: 768px) {
    font-size: 15px;
  }
`;

const FilledContainer = styled(Container)`
  background-color: #e0e0e0;
`;

const OutlinedContainer = styled(Container)`
  border: 1px solid #ccc;
`;

const PillIcon = styled(FontAwesomeIcon)`
  margin: ${props => (props.icon.iconName === 'times' ? '0 6.75px' : '0 4px')};
`;

const ResourceIconDiv = styled.div`
  padding-left: 5px;
`;

export const TEST_IDS = {
  pillIcon: 'ui-pill-icon',
  pill: 'ui-pill',
};
/**
 * Pill Component
 * @param {Object} props
 * @param {Object} props.label the text inside the pill
 * @param {Object} props.otherIcon if passed, then the corresponding icon will show up to the left of the text
 * @param {Object} props.resourceType if a resource type is passed, then adds resourceType icon to the left of text
 * @param {Object} props.icon *optional* a react compnent that will render to the left of the text
 * @param {Object} props.variant *defaults to filled* the render style [outlined, filled]
 * @param {Object} props.onDelete called if deletable is true, receives label as an arg
 * @param {Object} props.onClick click handler for entire pill if passed in
 */
export const Pill = ({
  label,
  otherIcon,
  resourceType,
  icon,
  variant,
  onDelete,
  onClick,
  deletable,
  type,
  ...rest
}) => {
  const props = {
    'data-testid': TEST_IDS.pill,
  };

  if (onClick) {
    props.onClick = onClick;
  }
  let specialIcon;
  if (otherIcon === 'topic') {
    specialIcon = faLayerGroup;
  } else if (otherIcon === 'journey') {
    specialIcon = faProjectDiagram;
  } else if (otherIcon === 'label') {
    specialIcon = faTag;
  } else if (otherIcon === 'persona') {
    specialIcon = faUser;
  } else if (otherIcon === 'search') {
    specialIcon = faSearch;
  }

  if (variant === variants.filled) {
    return (
      <FilledContainer {...props} {...rest} data-variant={variant.filled}>
        {resourceType && (
          <ResourceIconDiv>
            <ResourceTypeIcon type={resourceType} />
          </ResourceIconDiv>
        )}
        {otherIcon && <PillIcon icon={specialIcon} />}
        <span>{label}</span>
        {deletable && (
          <PillIcon
            aria-label="delete"
            data-testid={TEST_IDS.pillIcon}
            icon={faTimesCircle}
            onClick={() => onDelete(label)}
          />
        )}
      </FilledContainer>
    );
  } else if (variant === variants.outlined) {
    return (
      <OutlinedContainer {...props} {...rest} data-variant={variant.outlined}>
        {resourceType && (
          <ResourceIconDiv>
            <ResourceTypeIcon type={resourceType} />
          </ResourceIconDiv>
        )}
        {otherIcon && <PillIcon icon={specialIcon} />}
        <span>{label}</span>
        {deletable && (
          <PillIcon aria-label="delete" icon={faTimes} onClick={() => onDelete(label)} />
        )}
      </OutlinedContainer>
    );
  }
  return (
    <Container {...props} {...rest}>
      <span>{label}</span>
      {deletable && <PillIcon aria-label="delete" icon={faTimes} onClick={() => onDelete(label)} />}
    </Container>
  );
};

Pill.propTypes = {
  label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  icon: PropTypes.node,
  otherIcon: PropTypes.string,
  resourceType: PropTypes.string,
  variant: PropTypes.oneOf(Object.keys(variants)),
  onDelete: PropTypes.func,
  onClick: PropTypes.func,
  deletable: PropTypes.bool.isRequired,
};

Pill.defaultProps = {
  variant: variants.filled,
  deletable: true,
};

export default Pill;
