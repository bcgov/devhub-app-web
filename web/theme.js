/*
Copyright 2019 Province of British Columbia

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at 

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Created by Patrick Simonian
*/

/*
Emotion theming object

We are transitioning from common elements such as colors, background colors being abstracted away into
a themes provider. This is a very common practice now a days and makes it maintain themes on the fly
docs on theming is here https://emotion.sh/docs/theming#___gatsby

Docs on emotion https://emotion.sh/docs/introduction

--- our idea is to introduce emotion early and slowly replace css modules with emotion styled components
when and as needed. This is a slow processes but will lead to better and more maintaible styling in the long run.
*/

import { SPACING, DEVHUB_PALETTE } from './src/constants/designTokens';
export default {
  colors: {
    ...DEVHUB_PALETTE,
  },
  padding: {
    main: {
      mobile: `${SPACING['2x']} ${SPACING['2x']}`,
      desktop: `${SPACING['2x']} ${SPACING['13x']}`,
    },
  },
  breakpoints: {
    // specific breakpoints
    main: {
      desktop: '@media screen and (min-width: 932px)',
    },
  },
};
