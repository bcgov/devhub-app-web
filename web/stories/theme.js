import { create } from '@storybook/theming';
import { DEVHUB_PALETTE } from '../theme';

export default create({
  base: 'light',

  colorPrimary: 'hotpink',
  colorSecondary: 'deepskyblue',

  // UI
  appBg: 'white',
  appContentBg: 'white',
  appBorderColor: 'grey',
  appBorderRadius: 4,

  // Typography
  fontBase: 'sans-serif',
  fontCode: 'monospace',

  // Text colors
  textColor: DEVHUB_PALETTE.darkgrey,
  textInverseColor: 'rgba(255,255,255,0.9)',

  // Toolbar default and active colors
  barTextColor: DEVHUB_PALETTE.white,
  barSelectedColor: 'black',
  barBg: DEVHUB_PALETTE.blue,

  // Form colors
  inputBg: 'white',
  inputBorder: 'silver',
  inputTextColor: 'black',
  inputBorderRadius: 4,

  brandTitle: 'My custom storybook',
  brandImage:
    'https://u58tqm22fk-flywheel.netdna-ssl.com/wp-content/uploads/2016/06/BC-Gov-Logo-CMYK_pos.png',
});
