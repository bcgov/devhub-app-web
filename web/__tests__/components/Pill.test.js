import React from 'react';
import Pill, { TEST_IDS } from '../../src/components/UI/Pill';
import { render, cleanup, fireEvent } from '@testing-library/react';
import { ThemeProvider } from '@emotion/react';
import theme from '../../theme';
describe('Pill Component', () => {
  const props = {
    label: 'hello world',
    onClick: jest.fn(),
    onDelete: jest.fn(),
  };

  beforeEach(() => {
    props.onClick.mockReset();
    props.onDelete.mockReset();
  });

  afterEach(cleanup);

  it('renders a pill', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Pill {...props} />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('does not show font awesome icon if deletable is false', () => {
    const { queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <Pill {...props} deletable={false} />
      </ThemeProvider>,
    );
    expect(queryByTestId(TEST_IDS.pillIcon)).not.toBeInTheDocument();
  });

  it('calls onDelete if deletable and icon is clicked', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Pill {...props} />
      </ThemeProvider>,
    );
    const pillIcon = getByTestId(TEST_IDS.pillIcon);
    fireEvent.click(pillIcon);
    expect(props.onDelete).toHaveBeenCalledWith(props.label);
  });

  it('calls onClick when Pill is clicked and it exists in props', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Pill {...props} />
      </ThemeProvider>,
    );
    const pill = getByTestId(TEST_IDS.pill);
    fireEvent.click(pill);
    expect(props.onClick).toHaveBeenCalled();
  });

  it('does not call onClick when Pill is clicked and it does not exist in props', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Pill {...props} onClick={undefined} />
      </ThemeProvider>,
    );
    const pill = getByTestId(TEST_IDS.pill);
    fireEvent.click(pill);
    expect(props.onClick).not.toHaveBeenCalled();
  });
});
