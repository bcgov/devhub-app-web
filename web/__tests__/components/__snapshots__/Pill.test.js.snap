// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pill Component renders a pill 1`] = `
.emotion-2 {
  padding: 3px;
  border-radius: 1em;
  color: rgba(0,0,0,0.87);
  margin: 4px 5px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-color: transparent;
  border-width: 1px;
  border-style: solid;
  background-color: #e0e0e0;
}

.emotion-2:focus {
  outline: none;
}

.emotion-2 > span {
  margin: 0 4px;
  font-size: 0.875em;
}

@media only screen and (max-width:768px) {
  .emotion-2 {
    font-size: 15px;
  }
}

.emotion-0 {
  margin: 0 4px;
}

<button
  class="emotion-2 emotion-3"
  data-testid="ui-pill"
>
  <span>
    hello world
  </span>
  <svg
    aria-hidden="true"
    aria-label="delete"
    class="svg-inline--fa fa-times-circle fa-w-16  emotion-0 emotion-1"
    data-icon="times-circle"
    data-prefix="fas"
    data-testid="ui-pill-icon"
    focusable="false"
    role="img"
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
      fill="currentColor"
    />
  </svg>
</button>
`;
